libraries/ms-drivers/src/ltc_afe_impl.c

```diff
@@ -0,0 +1,302 @@
+#include "ltc_afe_impl.h"
+#include <stddef.h>
+#include <string.h>
+#include "crc15.h"
+#include "delay.h"
+#include "log.h"
+#include "ltc6811.h"
+
+// - 12-bit, 16-bit and 24-bit values are little endian
+// - commands and PEC are big endian
+
+static uint16_t s_read_reg_cmd[NUM_LTC_AFE_REGISTERS] = {
+  LTC6811_RDCFG_RESERVED,  LTC6811_RDCVA_RESERVED,   LTC6811_RDCVB_RESERVED,
+  LTC6811_RDCVC_RESERVED,  LTC6811_RDCVD_RESERVED,   LTC6811_RDAUXA_RESERVED,
+  LTC6811_RDAUXA_RESERVED, LTC6811_RDSTATA_RESERVED, LTC6811_RDSTATB_RESERVED,
+  LTC6811_RDCOMM_RESERVED
+};
+
+static uint8_t s_voltage_reg[NUM_LTC_AFE_VOLTAGE_REGISTERS] = {
+  LTC_AFE_REGISTER_CELL_VOLTAGE_A,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_B,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_C,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_D,
+};
+
+static void prv_wakeup_idle(LtcAfeStorage *afe) {
+  LtcAfeSettings *settings = &afe->settings;
+  // Wakeup method 2 - pair of long -1, +1 for each device
+  for (size_t i = 0; i < settings->num_devices; i++) {
+    gpio_set_state(&settings->cs, GPIO_STATE_LOW);
+    gpio_set_state(&settings->cs, GPIO_STATE_HIGH);
+    // Wait for 300us - greater than tWAKE, less than tIDLE
+    delay_us(300);
+  }
+}
+
+static StatusCode prv_build_cmd(uint16_t command, uint8_t *cmd, size_t len) {
+  if (len != 4) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  cmd[0] = (uint8_t)(command >> 8);
+  cmd[1] = (uint8_t)(command & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec);
+
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_read_register(LtcAfeStorage *afe, LtcAfeRegister reg, uint8_t *data,
+                                    size_t len) {
+  if (reg > NUM_LTC_AFE_REGISTERS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint16_t reg_cmd = s_read_reg_cmd[reg];
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(reg_cmd, cmd, SIZEOF_ARRAY(cmd));
+
+  prv_wakeup_idle(afe);
+  return spi_exchange(afe->settings.spi_port, cmd, 4, data, len);
+}
+
+// read from a voltage register
+static StatusCode prv_read_voltage(LtcAfeStorage *afe, LtcAfeVoltageRegister reg,
+                                   LtcAfeVoltageRegisterGroup *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTERS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  size_t len = sizeof(LtcAfeVoltageRegisterGroup) * afe->settings.num_devices;
+  return prv_read_register(afe, s_voltage_reg[reg], (uint8_t *)data, len);
+}
+
+// start cell voltage conversion
+static StatusCode prv_trigger_adc_conversion(LtcAfeStorage *afe) {
+  LtcAfeSettings *settings = &afe->settings;
+  uint8_t mode = (uint8_t)((settings->adc_mode + 1) % 3);
```

Comment:
any chance you could provide a comment for this? seems a bit funky but I'm ok with it

Reply:
Whelp, I'll look into it I guess :)

Reply:
Actually it's probably fine, as long as no one looks at it haha. Pretty sure it's just whatever conversion necessary to get enum vals to output values.